### ISSUE TITLE
Feature Request: Connection URL for Postgres

### ISSUE THREAD

**Issue by drbobbeaty** [hankinsoft/SQLPro#810]
In my use-case, I don't have the typical credentials of username/password/host - but I have  URL that contains all that information encoded in it, and `psql` can connect with it, but SQLPro can't.

Would it be possible to add this capability to SQLPro for Postgres?

**Comment by hankinsoft** [comment]
Thats something I can take a look into adding - I have this option for Oracle, so it shouldn't be overly difficult. That said a manually specified connection string would NOT work for any SSL operations or file based options. No eta, but I'll take a look into this hopefully within the next few weeks.

**Comment by hankinsoft** [comment]
Can you try https://d3fwkemdw8spx3.cloudfront.net/postgres/SQLProPostgres.2022.70.app.zip? At the bottom right when editing a connection choose 'Advanced' -> 'Connection string'. Please let me know if this works as you would expect.

**Comment by drbobbeaty** [comment]
Works great! Just as I'd hoped.

Will this make its way to the iPadOS app soon? (that's where I'd use it most)

**Comment by hankinsoft** [comment]
Sure, I'll take a look into adding it for iOS. Part of the code is shared, so I just need to do the UI portion. Just to confirm, you are using SQLPro for Postgres and not SQLPro Studio? (That lets me know which version I need to update first).

**Comment by drbobbeaty** [comment]
SQLPro for Postgres

**Comment by hankinsoft** [comment]
Hi. The lastest SQLPro for Postgres on iOS allows you to specify a connection string for postgres. If you get the chance, please give it a try and let me know if it works as expected.

**Comment by drbobbeaty** [comment]
@hankinsoft I've just tried it, and it's not going _super well_. 🙂

First, when I go to an existing connection that mistakenly had the Connection String in the "Host/IP" slot, I can't edit that field. I was hoping to copy it out of there and into the Connection String field.

So I fell back to another source of the URL, and tried that. I was able to put the Connection String in the little pop-up modal, and then the app updated the Host/IP and such - which made me smile, but then it wouldn't connect... When I went back in to the Connection settings, the Connection String was back in the Host/IP field. 🤔

So I deleted that Connection and made a new one - no big deal. I put in the Connection String, and it pulled out the host and all that, but then I tried to save it with an Alias, and it crashed.

So I repeated it, and I now can't seem to save a new connection, created with a Connection String, and an Alias properly.

Even when I skip the Alias, and just enter the Connection String and "Save", I go to connect and it's not connecting. This all was great to the same Connection String in the macOS App that started this thread - so I know the DB is OK.

In short...

* I can't get connected no matter what I try - old connection or new
* I can't seem to save a new Connection with a Connection String and Alias.
* I can't seem to save the password for the Connection String, as it's asking me for it when I try to connect, and when I enter it, it fails... The message is: "Connection to server on socket "/tmp/.s.PGSQL.5432" failed: No such file or directory" - but it worked on the macOS app.

So I'm kinda stuck. 🙁

Any ideas are welcome.

**Comment by drbobbeaty** [comment]
@hankinsoft More interesting debugging... When I create a new Connection, and put in the Connection String, and then - from the same dialog, hit "Test Connection" - that Works! But if I add an Alias, and then hit "Save & Connect" it asks me for a Password - and I know this one is included in the URL... So I can't connect.

Then, when I repeat this and just hit "Save" - it doesn't save the Connection String or the Alias.

So just those are something to look at... because it's 100% repeatable for me with several different Connection Strings.

**Comment by drbobbeaty** [comment]
@hankinsoft OK... I found the "Show Logs" in the app - very handy. Here is the section of the logs starting from when I put in the Connection String and it's properly parsed, and tested, and then the failure to save:

```
2022/11/03 10:37:42:938  SQLProPasscodeManager errSecItemNotFound (getPasscode)
2022/11/03 10:37:42:939  doGetSubscriptionActive entered
2022/11/03 10:37:42:947  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/03 10:37:42:947  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/03 10:37:42:947  Is lifetime.
2022/11/03 10:37:42:947  maxExpirationDate is lifetime.
2022/11/03 10:37:42:947  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/03 10:37:42:948  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
2022/11/03 10:37:42:948  doGetSubscriptionActive entered
2022/11/03 10:37:42:950  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/03 10:37:42:950  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/03 10:37:42:950  Is lifetime.
2022/11/03 10:37:42:950  maxExpirationDate is lifetime.
2022/11/03 10:37:42:950  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/03 10:37:42:950  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
2022/11/03 10:38:04:659  doConnectToConnectionModel: going to connect (postgres.libcloud.platter.app:5678)
2022/11/03 10:38:04:659  PostgresDatabaseClient connect called (PQlibVersion: v140005).
2022/11/03 10:38:05:107  checkSSHFingerprintError returning NO
2022/11/03 10:38:10:743  doRefreshConnectionsReloadOnMain started
2022/11/03 10:38:10:743  doRefreshConnectionsReloadOnMain finished
2022/11/03 10:38:10:880  doRefreshConnectionsReloadOnMain started
2022/11/03 10:38:10:880  doRefreshConnectionsReloadOnMain finished
2022/11/03 10:38:13:708  doGetSubscriptionActive entered
2022/11/03 10:38:13:714  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/03 10:38:13:714  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/03 10:38:13:715  Is lifetime.
2022/11/03 10:38:13:715  maxExpirationDate is lifetime.
2022/11/03 10:38:13:715  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/03 10:38:13:715  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
2022/11/03 10:38:13:729  SQLProPasscodeManager errSecItemNotFound (getPasscode)
2022/11/03 10:38:13:741  doGetSubscriptionActive entered
2022/11/03 10:38:13:743  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/03 10:38:13:743  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/03 10:38:13:743  Is lifetime.
2022/11/03 10:38:13:743  maxExpirationDate is lifetime.
2022/11/03 10:38:13:743  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/03 10:38:13:743  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
```

I think the key line is:

```
2022/11/03 10:38:05:107  checkSSHFingerprintError returning NO
```

and I know if I just use:

```bash
$ psql <url>
```

it connects just fine... so I'm not sure.

-----

But I tried another Connection String to a different database, and that log looks like:

```
2022/11/03 10:45:23:831  SQLProPasscodeManager errSecItemNotFound (getPasscode)
2022/11/03 10:45:23:832  doGetSubscriptionActive entered
2022/11/03 10:45:23:841  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/03 10:45:23:841  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/03 10:45:23:841  Is lifetime.
2022/11/03 10:45:23:841  maxExpirationDate is lifetime.
2022/11/03 10:45:23:841  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/03 10:45:23:841  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
2022/11/03 10:45:23:842  doGetSubscriptionActive entered
2022/11/03 10:45:23:843  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/03 10:45:23:843  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/03 10:45:23:843  Is lifetime.
2022/11/03 10:45:23:843  maxExpirationDate is lifetime.
2022/11/03 10:45:23:843  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/03 10:45:23:844  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
2022/11/03 10:45:52:940  doConnectToConnectionModel: going to connect (otto.db.elephantsql.com:5432)
2022/11/03 10:45:52:940  PostgresDatabaseClient connect called (PQlibVersion: v140005).
2022/11/03 10:45:53:485  checkSSHFingerprintError returning NO
2022/11/03 10:45:58:221  doRefreshConnectionsReloadOnMain started
2022/11/03 10:45:58:221  doRefreshConnectionsReloadOnMain finished
2022/11/03 10:45:58:362  doRefreshConnectionsReloadOnMain started
2022/11/03 10:45:58:363  doRefreshConnectionsReloadOnMain finished
2022/11/03 10:46:01:622  doGetSubscriptionActive entered
2022/11/03 10:46:01:629  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/03 10:46:01:629  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/03 10:46:01:629  Is lifetime.
2022/11/03 10:46:01:629  maxExpirationDate is lifetime.
2022/11/03 10:46:01:629  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/03 10:46:01:629  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
2022/11/03 10:46:01:642  SQLProPasscodeManager errSecItemNotFound (getPasscode)
2022/11/03 10:46:01:654  doGetSubscriptionActive entered
2022/11/03 10:46:01:655  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/03 10:46:01:656  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/03 10:46:01:656  Is lifetime.
2022/11/03 10:46:01:656  maxExpirationDate is lifetime.
2022/11/03 10:46:01:656  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/03 10:46:01:656  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
```

and again we have the key, I feel:

```
2022/11/03 10:45:53:485  checkSSHFingerprintError returning NO
```

But again, the "Test" button works, and says all is Good. But then "Save & Connect" doesn't do either.

**Comment by drbobbeaty** [comment]
@hankinsoft with the latest update this morning, I can save the connection, and it works!

But I can't save the password... that toggle is greyed out on me. Is there a way to correct that? Something I'm missing?

**Comment by hankinsoft** [comment]
> with the latest update this morning, I can save the connection, and it works!
Awesome, great to hear it.

> But I can't save the password... that toggle is greyed out on me. Is there a way to correct that? Something I'm missing?
This is the one thing I wasn't sure about via your previous messages.

When a connection string is used the password has to come from the connection string.

This is the same on macOS as well:

<img width="499" alt="Screenshot 2022-11-04 at 9 43 57 AM" src="https://user-images.githubusercontent.com/3160528/199976153-45132f6b-1d75-43cc-a99c-b48936228116.png">

**Comment by drbobbeaty** [comment]
OK... fair enough. 👍

**Comment by hankinsoft** [comment]
So youve got the password in the connection string, but are still being prompted to enter it? Yes, please provide the log file.

**Comment by drbobbeaty** [comment]
@hankinsoft OK... here's what I did, and saw, and then the log.

* I activated the app, and went to Edit the Connection.
* I can see the password **_is_** in the URL, but it's not in the Password field in the Connection dialog.
* I click on Connection String, and then Save, and the password populates the Password field.
* I hit "Save and Connect" and get it asking me for the password.

Log here: 

```
2022/11/04 17:50:45:461  doGetSubscriptionActive entered
2022/11/04 17:50:45:468  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/04 17:50:45:468  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/04 17:50:45:468  Is lifetime.
2022/11/04 17:50:45:468  maxExpirationDate is lifetime.
2022/11/04 17:50:45:468  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/04 17:50:45:468  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
2022/11/04 17:50:46:086  Trait colleciton changed
2022/11/04 17:50:54:790  SQLProPasscodeManager errSecItemNotFound (getPasscode)
2022/11/04 17:50:54:804  doGetSubscriptionActive entered
2022/11/04 17:50:54:807  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/04 17:50:54:807  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/04 17:50:54:807  Is lifetime.
2022/11/04 17:50:54:807  maxExpirationDate is lifetime.
2022/11/04 17:50:54:807  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/04 17:50:54:807  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
2022/11/04 17:52:05:293  SQLProPasscodeManager errSecItemNotFound (getPasscode)
2022/11/04 17:52:05:293  doGetSubscriptionActive entered
2022/11/04 17:52:05:301  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/04 17:52:05:301  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/04 17:52:05:302  Is lifetime.
2022/11/04 17:52:05:302  maxExpirationDate is lifetime.
2022/11/04 17:52:05:302  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/04 17:52:05:302  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
2022/11/04 17:52:05:302  doGetSubscriptionActive entered
2022/11/04 17:52:05:304  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/04 17:52:05:304  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/04 17:52:05:304  Is lifetime.
2022/11/04 17:52:05:304  maxExpirationDate is lifetime.
2022/11/04 17:52:05:304  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/04 17:52:05:305  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
2022/11/04 17:54:53:242  doGetSubscriptionActive entered
2022/11/04 17:54:53:249  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/04 17:54:53:249  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/04 17:54:53:249  Is lifetime.
2022/11/04 17:54:53:249  maxExpirationDate is lifetime.
2022/11/04 17:54:53:249  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/04 17:54:53:249  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
2022/11/04 17:54:53:268  SQLProPasscodeManager errSecItemNotFound (getPasscode)
2022/11/04 17:54:53:280  doGetSubscriptionActive entered
2022/11/04 17:54:53:282  doGetSubscriptionActive going to enumerate 1 receipts
2022/11/04 17:54:53:282  Product (sqlpro.postgres.ios.lifetime) cancelled: (no), expires: <null>, origional purchase: 2022-11-03 08:49:10 +0000
2022/11/04 17:54:53:282  Is lifetime.
2022/11/04 17:54:53:282  maxExpirationDate is lifetime.
2022/11/04 17:54:53:282  Max expiration date [sqlpro.postgres.ios.lifetime] is (null).
2022/11/04 17:54:53:282  HSAppStoreSubscriptionHelper:getSubscriptionActive - Setapp license is not enabled
```

**Comment by hankinsoft** [comment]
If you just hit 'Save', then connect from the main screen, does it still prompt you for the password?

**Comment by drbobbeaty** [comment]
Yes. I pull up the Connection dialog to edit - hit the Connection String because the password field is empty, and then Save that to fill it again, and then Save that Connection... and then tap on the Connection, and it asks for the password.

I wonder if it's not saving the password? Because in the Connection Dialog, I can Test the connection, and it works.

**Comment by hankinsoft** [comment]
There have been some updates on saving connections. Could you update to the latest version and let me know if this seems resolved.

**Comment by drbobbeaty** [comment]
So far - and it's only been a few days, but _so far_ it seems the connections are being saved. Thanks! 👍

**Comment by hankinsoft** [comment]
Great! I'm going to mark it as closed, but if you run into it again, please feel free to re-open.
